### PR TITLE
Gb/bespoke sc

### PR DESCRIPTION
**Pull request by grantbuster**
This needs to be tested full scale but it won't break anything so opening a PR now

### CODE REVIEW

**Comment by ppinchuk** [comment]
Full scale test failed... I think the bespoke meta is missing the `sc_(row,col)_ind` columns: https://github.com/NREL/reV/blob/a947b6ba8677224a939a1b2ac9b97d91525f06b2/reV/bespoke/bespoke.py#L355

Will add a review after patch

**Comment by grantbuster** [comment]
There's actually kind of a big issue that we need to make sure we're calculating LCOE(FCR) for the timeseries SAM simulation based on inputs from the optimized plant layout... Need to figure out with PJ how to link up the cost parameters...

**Comment by ppinchuk** [comment]
Yea, Owen, Travis, and I actually already talked about that. I was going to patch it in with a more general reV maintenance patch sometime this week

**Comment by grantbuster** [comment]
Yeah here's what I'm thinking: FCR and VOP would be taken from the input config and would be fixed, cap cost and fixed op cost would be based on the optimized plant. I'll just add optional inputs (required if `lcoe_fcr` is requested) to the `sam_sys_inputs` that would be `capital_cost_per_mw` and `fixed_operating_cost_per_mw` and then have the user input those.

**Comment by ppinchuk** [comment]
Couldn't you just compute `capital_cost_per_mw ` and `fixed_operating_cost_per_mw` from the input config (before running the optimization)? `capital_cost`, `fixed_operating_cost`, and `system_capacity` are already required parameters

**Comment by grantbuster** [comment]
Yep we could do that too! I'm not sure if system_capacity is really required for bespoke but I agree that makes more sense than new keys

**Comment by ppinchuk** [comment]
Ah fair. We have been throwing it in anyways, probably because old habits die hard XD

**Comment by grantbuster** [comment]
Good idea though, i'll go ahead and implement that

File: reV/bespoke/cli_bespoke.py [hunk 1]
```diff
@@ -274,7 +297,10 @@ def valid_config_keys():
               'the turbine rotor diameter.')
 @click.option('--ga_kwargs', '-gakw', type=STR, default=None,
               show_default=True,
-              help='Dictionary of keyword arguments to pass to GA initialization.')
+              help='Dictionary of keyword arguments to pass to GA '
+              'initialization.If `None`, default initialization values are '
+              'used. See :class:`~reV.bespoke.gradient_free.GeneticAlgorithm` '
```
Comment (ppinchuk) [inline reV/bespoke/cli_bespoke.py:302]:
```suggestion
              'used. See :class:`~reV.bespoke.gradient_free.GeneticAlgorithm` '
```
```suggestion
              'used. See reV.bespoke.gradient_free.GeneticAlgorithm '
```
My bad here - the tags I threw in front of that were for sphinx documentation purposes only and probably aren't too useful for a cli user :p

File: reV/bespoke/bespoke.py [hunk 2]
```diff
@@ -521,14 +568,78 @@ def plant_optimizer(self):
...
+        cap_cost = self.plant_optimizer.capital_cost
+        foc = self.plant_optimizer.fixed_operating_cost
+        voc = self.plant_optimizer.variable_operating_cost
+
+        lcoe_kwargs = {}
+        lcoe_kwargs['fixed_charge_rate'] = fcr
+        lcoe_kwargs['system_capacity'] = sys_cap
+        lcoe_kwargs['capital_cost'] = cap_cost
+        lcoe_kwargs['fixed_operating_cost'] = foc
+        lcoe_kwargs['variable_operating_cost'] = voc
```
Comment (ppinchuk) [inline reV/bespoke/bespoke.py:636]: Any reason not to just build a dictionary right away?
```suggestion
        lcoe_kwargs['variable_operating_cost'] = voc
```
```suggestion
        lcoe_kwargs = {
            'fixed_charge_rate': self.original_sam_sys_inputs['fixed_charge_rate']
            'system_capacity': self.plant_optimizer.capacity
            'capital_cost': self.plant_optimizer.capital_cost
            'fixed_operating_cost': self.plant_optimizer.fixed_operating_cost
            'variable_operating_cost': self.plant_optimizer.variable_operating_cost
        }
```

Comment (ppinchuk) [inline reV/bespoke/bespoke.py:636]: Any reason not to just build a dictionary right away?
```suggestion
        lcoe_kwargs['variable_operating_cost'] = voc
```
```suggestion
        lcoe_kwargs = {
            'fixed_charge_rate': self.original_sam_sys_inputs['fixed_charge_rate']
            'system_capacity': self.plant_optimizer.capacity
            'capital_cost': self.plant_optimizer.capital_cost
            'fixed_operating_cost': self.plant_optimizer.fixed_operating_cost
            'variable_operating_cost': self.plant_optimizer.variable_operating_cost
        }
```

File: reV/bespoke/bespoke.py [hunk 3]
```diff
@@ -521,14 +568,78 @@ def plant_optimizer(self):
...
+
+        for k, v in lcoe_kwargs.items():
+            self._meta[k] = v
```
Comment (ppinchuk) [inline reV/bespoke/bespoke.py:639]: Should we document this side-effect? It does mutate the object in an important way

Comment (ppinchuk) [inline reV/bespoke/bespoke.py:639]: Should we document this side-effect? It does mutate the object in an important way

File: reV/bespoke/bespoke.py [hunk 4]
```diff
@@ -796,17 +928,28 @@ def __init__(self, excl_fpath, res_fpath, tm_dset,
...
+            capital cost in $. Has access to the same variables as the
+            objective_function.
+        fixed_operating_cost_function : str
+            The plant annual fixed operating cost function as a string, must
+            return the fixed operating cost in $/year. Has access to the same
+            variables as the objective_function.
+        variable_operating_cost_function : str
+            The plant annual variable operating cost function as a string, must
+            return the variable operating cost in $/kWh. Has access to the same
+            variables as the objective_function.
```
Comment (ppinchuk) [inline reV/bespoke/bespoke.py:952]: Excellent docstrings. As an end user, I am especially happy about the unit specifications.

Comment (ppinchuk) [inline reV/bespoke/bespoke.py:952]: Excellent docstrings. As an end user, I am especially happy about the unit specifications.

File: reV/bespoke/place_turbines.py [hunk 5]
```diff
@@ -267,24 +298,31 @@ def nturbs(self):
...
+        """This is the final optimized plant nameplate capacity (kW)"""
         if self.optimized_design_variables is not None:
             return self.turbine_capacity * self.nturbs
         else:
             return None
 
     @property
     def area(self):
-        """This is the area available for wind turbine placement"""
+        """This is the area available for wind turbine placement (km2)"""
```
Comment (ppinchuk) [inline reV/bespoke/place_turbines.py:309]: Again, definitely appreciate the unit clarifications :)

Comment (ppinchuk) [inline reV/bespoke/place_turbines.py:309]: Again, definitely appreciate the unit clarifications :)

**Review by ppinchuk — APPROVED** [review]
Generally, looks great! Left a few minor comments but you can ignore them.
Full conus run is successful with this branch: `/shared-projects/rev/projects/weto/fy22/bespoke/rev/test_runs_conus_sc_new_cost_funcs`

File: reV/bespoke/bespoke.py (hunk 3 above)
Comment (grantbuster) [reply reV/bespoke/bespoke.py:639]: I think it's fine, the purpose here is just to add more fields to the meta data

File: reV/bespoke/bespoke.py (hunk 4 above)
Comment (grantbuster) [reply reV/bespoke/bespoke.py:952]: ugh yeah units are so obnoxious in research software!

File: reV/bespoke/cli_bespoke.py (hunk 1 above)
Comment (grantbuster) [reply reV/bespoke/cli_bespoke.py]: fixed

File: reV/bespoke/bespoke.py (hunk 2 above)
Comment (grantbuster) [reply reV/bespoke/bespoke.py]: fixed

**Review by ppinchuk — APPROVED** [review]
LGTM

**Pull request merged**